sdk/storage/azure-storage-file-share/src/test/java/com/azure/storage/file/share/FileApiTests.java

```diff
@@ -637,6 +640,91 @@ public void downloadRetryDefault() {
         assertEquals(bodyStr, DATA.getDefaultText());
     }
 
+    @Test
+    public void downloadToFileWithPartialDownloads() throws Exception {
+        File uploadFile = File.createTempFile(CoreUtils.randomUuid().toString(), ".txt");
+        uploadFile.deleteOnExit();
+        Files.write(uploadFile.toPath(), DATA.getDefaultBytes());
+
+        File outFile = new File(generatePathName() + ".txt");
+        if (outFile.exists()) {
+            assertTrue(outFile.delete());
+        }
+
+        MockPartialResponsePolicy policy = new MockPartialResponsePolicy(5);
+
+        // Create a ShareFileClient for download using the custom pipeline
+        ShareFileClient downloadClient
+            = getFileClient(ENVIRONMENT.getPrimaryAccount().getCredential(), primaryFileClient.getFileUrl(), policy);
+
+        // Upload the test data
+        primaryFileClient.create(DATA.getDefaultDataSize());
+        primaryFileClient.uploadFromFile(uploadFile.toString());
+
+        downloadClient.downloadToFile(outFile.toString());
+
+        byte[] downloadedData = Files.readAllBytes(outFile.toPath());
+        Assertions.assertArrayEquals(DATA.getDefaultBytes(), downloadedData);
+
+        // Assert that we retried the correct number of times (5)
+        assertEquals(0, policy.getTriesRemaining());
+
+        // Assert that the range headers that were retried match what was returned from MockPartialResponsePolicy
+        List<int[]> actualRanges = policy.getRangesList();
+        for (int i = 0; i < actualRanges.size(); i++) {
+            int[] range = actualRanges.get(i);
+            assertEquals(i, range[0]);
+            assertEquals(DATA.getDefaultDataSize() - 1, range[1]);
+        }
+
+        // Clean up
+        Files.deleteIfExists(outFile.toPath());
+        Files.deleteIfExists(uploadFile.toPath());
+    }
+
+    @Test
+    public void downloadToFileRetryExhausted() throws Exception {
+        File uploadFile = File.createTempFile(CoreUtils.randomUuid().toString(), ".txt");
+        uploadFile.deleteOnExit();
+        Files.write(uploadFile.toPath(), DATA.getDefaultBytes());
+
+        File outFile = new File(generatePathName() + ".txt");
+        if (outFile.exists()) {
+            assertTrue(outFile.delete());
+        }
+
+        MockPartialResponsePolicy policy = new MockPartialResponsePolicy(6);
+
+        // Create a ShareFileClient for download using the custom pipeline
+        ShareFileClient downloadClient
+            = getFileClient(ENVIRONMENT.getPrimaryAccount().getCredential(), primaryFileClient.getFileUrl(), policy);
+
+        // Upload the test data
+        primaryFileClient.create(DATA.getDefaultDataSize());
+        primaryFileClient.uploadFromFile(uploadFile.toString());
+
+        try {
+            downloadClient.downloadToFile(outFile.toString());
+        } catch (Throwable e) {
+            assertInstanceOf(IOException.class, e.getCause());
+        }
```

Comment:
If we use assertThrows, the exception we get back is a ReactiveException from the Mono/Flux wrapper. It would look like something like this:

assertThrows(ReactiveException.class, e);

Would this be better?

Reply:
I'm a little surprised that the SDK would throw an exception specific to Reactor instead of the direct IOException. Maybe we can assert throws against the ReactiveException but save the return value from `assertThrows` and then assert that the return value's `getCause()` is an instance of an `IOException`? The main reason for the second assertion would be so that we are asserting that the root cause for the assertion was our IOException